wallet-core/src/eip2333.rs

```diff
@@ -0,0 +1,631 @@
+/// # Panics
+/// Panics if any of the `ikm_to_lamport_sk` fails.
+fn parent_sk_to_lamport_pk(parent_sk: &BlsScalar, index: u32) -> Vec<u8> {
+    // salt = I2OSP(index, 4)
+    let salt = index.to_be_bytes();
+
+    // IKM = I2OSP(parent_SK, 32)
+    let ikm = parent_sk.to_be_bytes();
+
+    // lamport_0 = IKM_to_lamport_SK(IKM, salt)
+    let mut lamport_0 = [[0u8; SHA256_DIGEST_SIZE]; HKDF_DIGESTS];
+    ikm_to_lamport_sk(ikm.as_slice(), salt.as_slice(), &mut lamport_0);
+
+    // not_IKM = flip_bits(IKM)
+    let not_ikm = ikm.map(|byte| !byte);
+
+    // lamport_1 = IKM_to_lamport_SK(not_IKM, salt)
+    let mut lamport_1 = [[0u8; SHA256_DIGEST_SIZE]; HKDF_DIGESTS];
+    ikm_to_lamport_sk(not_ikm.as_slice(), salt.as_slice(), &mut lamport_1);
+
+    // Combine `lamport_0` and `lamport_1` arrays
+    let mut lamport_combined = [[0u8; SHA256_DIGEST_SIZE]; HKDF_DIGESTS * 2];
+    lamport_combined[..HKDF_DIGESTS]
+        .clone_from_slice(&lamport_0[..HKDF_DIGESTS]);
+    lamport_combined[HKDF_DIGESTS..HKDF_DIGESTS * 2]
+        .clone_from_slice(&lamport_1[..HKDF_DIGESTS]);
+
+    // for i  in 1, .., 255
+    //    lamport_PK = lamport_PK | SHA256(lamport_0[i])
+    // for i  in 1, .., 255
+    //    lamport_PK = lamport_PK | SHA256(lamport_1[i])
+    let mut lamport_pk = [0u8; HKDF_OUTPUT_SIZE * 2];
+    for i in 0..HKDF_DIGESTS * 2 {
+        let sha_slice = &Sha256::digest(lamport_combined[i]);
+        lamport_pk[i * SHA256_DIGEST_SIZE..(i + 1) * SHA256_DIGEST_SIZE]
+            .clone_from_slice(sha_slice);
+    }
+
+    // compressed_lamport_PK = SHA256(lamport_PK)
+    Sha256::digest(lamport_pk).to_vec()
+}
+
+/// Derives a BLS scalar using HKDF extraction and expansion.
+///
+/// # EIP2333 Specification:
+///
+/// Inputs
+///  - `IKM`, a secret octet string >= 256 bits in length
+///  - `key_info`, an optional octet string (default="", the empty string)
+///
+/// Output
+///  - `SK`, the corresponding secret key, an integer 0 <= SK < r.
+///
+/// Definitions
+///  - `L`: integer given by `ceil((3 * ceil(log2(r)))` / 16).(L=48)
+///  - `r`: the order of the BLS 12-381 curve defined in the v4 draft IETF BLS
+///    signature scheme standard
+///
+/// Procedure
+///
+///  1. `salt = "BLS-SIG-KEYGEN-SALT-"`
+///  2. `SK = 0`
+///  3. `while SK == 0:`
+///  4. `salt = H(salt)`
+///  5. `PRK = HKDF-Extract(salt, IKM || I2OSP(0, 1))`
+///  6. `OKM = HKDF-Expand(PRK, key_info || I2OSP(L, 2), L)`
+///  7. `SK = OS2IP(OKM) mod r`
+///  8. `return SK`
+///
+///
+/// # Arguments
+/// * `ikm`      - The initial keying material for HKDF.
+/// * `key_info` - The info for HKDF.
+///
+/// # Returns
+/// The derived secret key.
+///
+/// # Panics
+/// Panics if HKDF extraction/expansion fail.
+#[allow(clippy::similar_names)]
+fn hkdf_mod_r(ikm: &[u8], key_info: &[u8]) -> BlsScalar {
+    // IKM || I2OSP(0, 1)
+    let ikm_combined = [ikm, &[0u8]].concat();
+    // key_info || I2OSP(L, 2)
+    let key_info_combined = [key_info, &[0u8, 48u8]].concat();
```

Comment:
Might be nice to define `L` at the beginnig of the function so that 48 doesn't always appear as a "magic number"

Reply:
It was. I think I removed it because of clippy complaining about it. I'll try to reintroduce it

Reply:
done